### DISCUSSION TITLE
Jest Testing Errors

### DISCUSSION THREAD

**Discussion by snarkipus** [eclipse-langium/langium#653] [Q&A] [answered]
I don't think this is a Langium issue necessarily. I previously narrowed it down to a unresolved issue with `jest-chain` ([#34](https://github.com/mattphillips/jest-chain/issues/34)) ... or at least I thought I had.

I'm not entirely sure when this broke (or how), but I'm constantly getting `TypeError: result.message is not a function` for failing tests in Jest.

I swear I resolved this once by simply rolling back the version of jest (like [this](https://github.com/mattphillips/jest-chain/issues/34#issuecomment-1153149680)), but that doesn't seem to work anymore and I narrowly avoided npm hell by trying. Is this something you all have observed?

**Comment by montymxb** [comment, marked as answer]
@snarkipus I've seen this pop up before on a separate project when working with `jest-expect-message`, but it's an elusive error. However, I haven't personally seen this show up in Langium myself. Is it possible something got stuck version-wise in **node_modules**?

**Reply by snarkipus** [reply]
@montymxb honestly, I have absolutely no idea how to even go about checking that.

**Reply by montymxb** [reply]
I would check the version of jest that's installed for Langium with `npm list jest`. Also is this off the main branch, or is this the one of the previous releases (like 0.4.0), might be interesting to know if this is something that we've since updated on the current main branch.

You mentioned things were working before, so you could try a clean install with `npm ci` and see if that corrects any inconsistencies in your node_modules. Sometimes if there's something off version-wise, that can fix it.

Personally I'm running with jest **28.1.3** for Langium and I haven't had any issues there. I've read that version 26 sometimes had issues with this, but not sure if that holds for newer versions as well. Depending on what version of jest you're using, you could also try a newer one, and see if that corrects it (you would also want to update **ts-jest**, and **@types/jest** as well in your package.json to keep them close in version)

Hopefully a clean install, or possibly tinkering with the version helps. If you're still stuck after that, let me know, and I can try to figure out something more specific to help out 👍 .

**Reply by snarkipus** [reply]
Thanks @montymxb - I updated jest to 28.1.3 and did a clean install to no avail. I'm working from the latest commit in the github repo. Here's my dev deps from the package.json and the `npm list` output just in case something jumps out at you. Thanks for looking into it, I really appreciate it.
```json
  "devDependencies": {
    "@types/jest": "^28.1.1",
    "@types/jest-expect-message": "^1.0.3",
    "@typescript-eslint/eslint-plugin": "^5.28.0",
    "@typescript-eslint/parser": "^5.28.0",
    "eslint": "^8.17.0",
    "eslint-plugin-header": "^3.1.1",
    "jest": "^28.1.1",
    "jest-expect-message": "^1.0.2",
    "shx": "^0.3.4",
    "ts-jest": "^28.0.5",
    "typescript": "^4.7.4"
  },
```
```term
langium-workspaces@ /home/snarkipus/Projects/langium/langium-core/langium
├── @types/jest-expect-message@1.0.3
├── @types/jest@28.1.8
├── @typescript-eslint/eslint-plugin@5.28.0
├── @typescript-eslint/parser@5.28.0
├── eslint-plugin-header@3.1.1
├── eslint@8.17.0
├─┬ generator-langium@0.4.0 -> ./packages/generator-langium
│ ├── @types/lodash@4.14.175
│ ├── @types/yeoman-generator@5.2.7
│ ├── chalk@4.1.2
│ ├── lodash@4.17.21
│ └── yeoman-generator@5.5.2
├── jest-expect-message@1.0.2
├── jest@28.1.3
├─┬ langium-arithmetics-dsl@0.4.0 -> ./examples/arithmetics
│ ├── chalk@4.1.2 deduped
│ ├── commander@8.2.0
│ ├── langium-cli@0.4.0 deduped -> ./packages/langium-cli
│ ├── langium@0.4.0 deduped -> ./packages/langium
│ ├── vscode-languageclient@8.0.1
│ └── vscode-languageserver@8.0.1
├─┬ langium-cli@0.4.0 -> ./packages/langium-cli
│ ├── @types/fs-extra@9.0.13
│ ├── chalk@4.1.2 deduped
│ ├── commander@7.2.0
│ ├── fs-extra@9.1.0
│ ├── jsonschema@1.4.0
│ ├── langium@0.4.0 deduped -> ./packages/langium
│ └── lodash@4.17.21 deduped
├─┬ langium-domainmodel-dsl@0.4.0 -> ./examples/domainmodel
│ ├── chalk@4.1.2 deduped
│ ├── commander@8.2.0 deduped
│ ├── langium-cli@0.4.0 deduped -> ./packages/langium-cli
│ ├── langium@0.4.0 deduped -> ./packages/langium
│ ├── lodash@4.17.21 deduped
│ ├── vscode-languageclient@8.0.1 deduped
│ └── vscode-languageserver@8.0.1 deduped
├─┬ langium-sprotty@0.4.0 -> ./packages/langium-sprotty
│ ├── langium@0.4.0 deduped -> ./packages/langium
│ └── sprotty-protocol@0.11.0
├─┬ langium-statemachine-dsl@0.4.0 -> ./examples/statemachine
│ ├── chalk@4.1.2 deduped
│ ├── commander@8.2.0 deduped
│ ├── langium-cli@0.4.0 deduped -> ./packages/langium-cli
│ ├── langium@0.4.0 deduped -> ./packages/langium
│ ├── lodash@4.17.21 deduped
│ ├── vscode-languageclient@8.0.1 deduped
│ └── vscode-languageserver@8.0.1 deduped
├─┬ langium-vscode@0.4.0 -> ./packages/langium-vscode
│ ├── @types/vscode@1.60.0
│ ├── esbuild@0.14.25
│ ├── langium@0.4.0 deduped -> ./packages/langium
│ ├── vscode-languageclient@8.0.1 deduped
│ └── vscode-languageserver@8.0.1 deduped
├─┬ langium@0.4.0 -> ./packages/langium
│ ├── chevrotain@9.1.0
│ ├── langium-cli@0.4.0 deduped -> ./packages/langium-cli
│ ├── vscode-languageserver-textdocument@1.0.5
│ ├── vscode-languageserver@8.0.1 deduped
│ └── vscode-uri@3.0.2
├── shx@0.3.4
├── ts-jest@28.0.8
└── typescript@4.7.4
```

**Reply by montymxb** [reply]
@snarkipus Hmm, interesting. Packages were a little different on my end for a clean install, but even matching them explicitly to yours does not replicate the issue 🤔 . 

The next thing that comes to mind is your version of node & npm that's installed. Langium expects npm in the range **7.X.X**, but in practice I've used  npm **8.11.0** without issue -- coupled with node **v16.15.1**. It's possible if you're running an older version of node that there might be something odd going on. If you are on an older version, perhaps bumping up a bit might rectify the problem.

**Reply by snarkipus** [reply]
Perhaps it's something unique to my environment somehow. I"m not entirely sure.

Doing a clean install (`npm ci`) after setting npm to 8.11.0 (`npm install -g npm@8.11.0`) and setting node to 16.15.1 (`nvm install 16.15.1`) didn't change the behavior. I even tried downgrading npm to ^7.7 (`npm install npm@7.7`), and that didn't work either.

Node.js may be ubiquitous, and it might be relatively easy to pick up, but the package management and ecosystem are somewhat cumbersome.  Regardless, I do appreciate you looking into it.

**Reply by snarkipus** [reply]
It's definitely unique to my environment somehow. It's repeatable for me against a dummy hello-world project without any langium dependencies at all. I haven't tested it against the langium environment yet, but downgrading jest and all the associated packages seems to have resolved it for my dummy build. Thanks again for taking a look though @montymxb!
```json
  "dependencies": {
    "@types/jest": "^29.0.0",
    "@types/jest-expect-message": "^1.0.4",
    "@typescript-eslint/eslint-plugin": "^5.28.0",
    "@typescript-eslint/parser": "^5.28.0",
    "eslint": "^8.17.0",
    "eslint-plugin-header": "^3.1.1",
    "jest": "^26.6.3",
    "jest-expect-message": "^1.0.2",
    "npm": "^7.24.2",
    "shx": "^0.3.4",
    "ts-jest": "^26.5.6",
    "ts-node": "^10.9.1",
    "typescript": "^4.7.4"
  }
```

**Reply by montymxb** [reply]
No problem! And Interesting, so dropping Jest back down a bit resolved it? Maybe there's something more going on at the low-level, hmm 🤔 . Either way, I'm glad to hear that you were able to resolve it shifting the versions around a bit! I'll keep this in mind, especially if it shows up somewhere else .
